libyara/modules/macho.c

```diff
@@ -0,0 +1,1180 @@
+
+1. Redistributions of source code must retain the above copyright notice, this
+list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+this list of conditions and the following disclaimer in the documentation and/or
+other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its contributors
+may be used to endorse or promote products derived from this software without
+specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include <yara/macho.h>
+#include <yara/endian.h>
+#include <yara/modules.h>
+#include <yara/mem.h>
+
+#define MODULE_NAME macho
+
+typedef struct _FAT_DATA {
+
+    // Single Mach-O file headers may be out of current block in fat Mach-O
+    // binaries. This structure stores auxiliary information for such cases.
+
+    size_t parsed_size; // Number of already parsed bytes
+    size_t file_count;  // Number of already parsed files
+
+    size_t count;    // Number of out of block offsets
+    size_t* offsets; // Mach-O files offsets
+
+} FAT_DATA;
+
+
+// Initialize FAT_MACHO_DATA structure.
+
+void macho_init_fat_data_storage(
+    FAT_DATA* storage)
+{
+  storage->parsed_size = 0;
+  storage->file_count = 0;
+  storage->count = 0;
+  storage->offsets = NULL;
+}
+
+
+// Check if file is for 32-bit architecture.
+
+int macho_is_32(uint8_t* magic)
+{
+  // Magic must be [CE]FAEDFE or FEEDFA[CE].
+  return magic[0] == 0xce || magic[3] == 0xce;
+}
+
+
+// Check if file is for big-endian architecture.
+
+int macho_is_big(uint8_t* magic)
+{
+  // Magic must be [FE]EDFACE or [FE]EDFACF.
+  return magic[0] == 0xfe;
+}
+
+
+// Check for Mach-O fat binary signature.
+
+int macho_is_fat_file_block(uint32_t* magic)
+{
+  return *magic == FAT_MAGIC || *magic == FAT_CIGAM;
+}
+
+
+// Check for Mach-O binary signatures.
+
+int macho_is_file_block(uint32_t* magic)
+{
+  return *magic == MH_MAGIC || *magic == MH_MAGIC_64 ||
+      *magic == MH_CIGAM || *magic == MH_CIGAM_64;
+}
+
+
+// Get length of string with limit (some strings are terminated by size limit).
+
+size_t macho_limited_strlen(char* string, size_t limit)
```

Comment:
but as long as it's from ISO C90, microsoft didn't had time yet to include it in their toolchains, so this strnlen will not work for mingw/msvc builds. But libc version will be more optimal, so it will be good to use it if available